enterprise/backend/src/metabase_enterprise/semantic_search/index.clj

```diff
@@ -213,33 +215,37 @@
         excluded-kw (fn [column] (keyword (str "excluded." (name column))))]
     (zipmap update-keys (map excluded-kw update-keys))))
 
-(defn- throw-if-max-pg-len
-  [resource-name msg]
-  (when (> (count resource-name) 63)
-    (throw (ex-info msg
-                    {:table-name resource-name
-                     :length (count resource-name)
-                     :limit 63}))))
+(defn hash-identifier-if-exceeds-pg-limit
+  "Sometimes we need to generate new table/index names, such as when forcing a new index despite no change in index parameters.
+  When we do so we need to be sure any index names do not exceed the postgres limit for names. This function will hash the identifier
+  if it exceeds the length, and will get a name like index_${sha1} instead.
+
+  Note: The index parameters will still be available in index_metadata"
+  [identifier]
+  (if (<= (count identifier) 63)
+    identifier
+    (let [hashed-name (str "index_" (buddy-codecs/bytes->hex (buddy-hash/sha1 identifier)))]
+      (log/warnf "Using hashed name for index table %s as original table name %s exceeded the maximum table name length" hashed-name identifier)
+      hashed-name)))
 
 (defn model-table-suffix
   "Returns a new suffix for a table name, based on current timestamp"
   []
   (mod (.toEpochSecond (t/offset-date-time)) 10000000))
 
 (defn model-table-name
-  "Returns the table name for a model."
+  "Returns a default table name for a model. If the table name would exceed the 63 byte postgres limit, a hashed name is preferred."
   [embedding-model]
   (let [{:keys [model-name provider vector-dimensions]} embedding-model
         provider-name (embedding/abbrev-provider-name provider)
         abbrev-model-name (embedding/abbrev-model-name model-name)
-        result (str "index_" provider-name "_" abbrev-model-name "_" vector-dimensions "_" (model-table-suffix))]
-    (throw-if-max-pg-len result "Table name exceeds PostgreSQL limit")
-    result))
+        ideal-table-name (str "index_" provider-name "_" abbrev-model-name "_" vector-dimensions)]
+    (hash-identifier-if-exceeds-pg-limit ideal-table-name)))
 
 (defn default-index
   "Returns the default index spec for a model."
-  [embedding-model]
-  (let [table-name (model-table-name embedding-model)]
+  [embedding-model & {:keys [table-name]}]
+  (let [table-name (or table-name (model-table-name embedding-model))]
```

Comment:
The random suffix no longer happens for each call (made tests etc quite complex to think about) - it is predictable as before, in `pgvector-api/initialize-index!` we now use this option to vary the name al-carte when doing a `:force-reset? true`.